### DISCUSSION TITLE
Search by scope

### DISCUSSION THREAD

**Discussion by jaulz** [LastDragon-ru/php-packages#170] [Ideas]
It would be nice if there was a way to specify a scope on a search property that would be used instead of using operators.

**Comment by LastDragon-ru** [comment]
Do you mean Eloquent Model scope or just a set of predefined operators?

**Reply by jaulz** [reply]
I mean Eloquent Model scopes. Something like this:
```graphql
extend type Team {
  tags(filters: TagFilters @searchBy): [Tag]
    @paginate
}

type TagFilters {
  name: String @searchByScope(name: "ofName")
}
```

And then the model:
```php
  public function scopeOfName(Builder $query, string $name): Builder
  {
    return $query->where('name', 'LIKE', sprintf('\%%s\%', $name));
  }
```

I know this is a simple example which could also be solved by today's features but the scope could also be more complex.

**Reply by LastDragon-ru** [reply]
> I know this is a simple example which could also be solved by today's features but the scope could also be more complex.

Yep, it can be done via a custom operator.

> I mean Eloquent Model scope

Well, I was thinking about it some time ago. And there are a few problems 

1) The concrete model is unknown on manipulation/generation phase, so we cannot determine which scopes are available (and thus generate enum/type).

2) Even if the model will be known, there seems no built-in way to get a list of scopes.

3) Scopes may have parameters, they make generation more difficult.

Maybe I will return to it after current/planned issues about refactoring (#151,  #152,  #153), but no promises (I'm using scopes rarely...).

**Reply by jaulz** [reply]
I thought we can just take the value as it is (string, integer etc.) and pass it to the scope method of the query builder as the first parameter. No need to introspect anything because the scope name is passed to the directive.

Custom operators have the downside that they are global and I just need it for one type.

**Reply by LastDragon-ru** [reply]
> I thought we can just take the value as it is (string, integer etc.) and ...

The main problem is the point (1). There are several ways how to assign Type to Model (directories, `@model`, maybe something else). Also, there is `@namespace` directive that seems may affect how the class resolves (I'm even not sure that this directive supported now...). 

> Custom operators have the downside that they are global and I just need it for one type.

Actually, you can use `Operator::isAvailable($source)` to limit the type. But I agree, this is more side effect rather than expected usage. 

https://github.com/LastDragon-ru/lara-asp/blob/31279fa71d0b53c22c3ec71ddb8b7189738265de/packages/graphql/src/Builder/Contracts/Operator.php#L19

Will be good to be able to add different operators directly in graphql schema. Current version allows to add `Object` operator (like relationship) to a type definition (or to a field). But I'm not sure how it should look for `Extra` operators to avoid any conflicts and confusions (with `Object` and [other types](https://github.com/LastDragon-ru/lara-asp/blob/31279fa71d0b53c22c3ec71ddb8b7189738265de/packages/graphql/src/SearchBy/Operators.php#L57-L61)).

**Reply by jaulz** [reply]
Hmm, I still don't understand why we need to access any model at generation phase. If you take my example:
```graphql
type TagFilters {
  name: String @searchByScope(name: "ofName")
}
```

I indicated that the value is a string and that the scope `ofName` should be used. Now, couldn't we simply use the existing builder? Something like this:

```
if ($node->hasDirective('searchByScope')) {
  $directive = $node->getDirective('searchByScope');
  $scopeName = $directive->getAttribute('ofName');
  $value = $node->getValue();
  
  $builder->{$scopeName}($value);
} else {
  // Cover the existing cases
}
```

At generation phase we can just leave the filters that are annotated with `@searchByScope` and leave the type as it is.

**Reply by LastDragon-ru** [reply]
> Hmm, I still don't understand why we need to access any model at generation phase

The whole idea of this library is to provide deterministic and error free schema :) In your example the client should know the scopes or it will get an error.

I feel it might be something like:

```graphl
items(
   where: {
     anyOf: [
        scope: { ofName: { name: "a" } }
        scope: { ofName: { name: "b" } }
     ],
   }
) { ... } 
```

**Reply by jaulz** [reply]
Okay, now I got your point 😊  Though, I must admit that I don't really like that approach since it exposes internal information. The user shouldn't know whether it's a scope or not. He should just pass a value to a field and the framework should work according to the definition of the developer (e.g. via operator or via `@searchByScope`).
